File: raspPi/src/nodes/BaseNode.cpp
#include "BaseNode.hpp"
#include "MirrorNode.hpp"
#ifndef WIN32
#include <strings.h>
#endif
#include <sstream>
#include <algorithm>

using namespace std;

// Register attributes.
Attribute flagsAttrib("flags");
Attribute typeAttrib("type");
Attribute infoAttrib("info");
Attribute optionsAttrib("options");

// Define flag names 
std::array< std::tuple<const string, FlagType>, (size_t)FlagBitType::numflags > BaseNode::flagNames =
{
    make_tuple("hide"       , FlagType::hide),
    make_tuple("readonly"   , FlagType::readonly),
    make_tuple("logging"    , FlagType::logging),
    make_tuple("persist"    , FlagType::persist),
    make_tuple("query"      , FlagType::query),
};


BaseNode::BaseNode(const std::string &nodeName, BaseNode* pParentNode):
    recentChanges(0),
    attributeChanges(0),
    touchedAttributes(0),
    name(nodeName),
    pParent(pParentNode),
    nodeFlag(FlagType::none)
{
    if (pParent)
        pParent->SetNodeChanged();
}

BaseNode::~BaseNode()
{
}

BaseNode* BaseNode::GetRoot() const
{
    BaseNode* pNode = const_cast<BaseNode*>(this);
    while (pNode->pParent)
    {
        pNode = pNode->pParent;
    }
    return pNode;
}

// Observer 
void BaseNode::Subscribe(NodeObserver* pObserver)
{
    // Check if the same observer is already subscribing.
    if ( std::find(subscribers.begin(), subscribers.end(), pObserver) != subscribers.end() )
        return;
    // Add observer to list of subscribers.
    subscribers.push_front(pObserver);

}

void BaseNode::UnSubscribe(NodeObserver* pObserver)
{
    std::remove(subscribers.begin(), subscribers.end(), pObserver);
}

void BaseNode::Notify(bool recursive)
{
    if (recentChanges && 0x10) // Is this node changed during last update.
    {
        for (auto pObserver : subscribers)
        {
            pObserver->Notify();
        }
        if (recursive)
        {
            // Notify all childs aswell
            for (const auto & child : children)
            {
                child->Notify(recursive);
            }
        }
    }
}

bool BaseNode::LinkAllMirrors()
{
    bool success = true;
    MirrorNode *pMirror = dynamic_cast<MirrorNode*>(this);
    if (pMirror != nullptr) // Check if this node is a mirror.
    {
        success &= pMirror->LinkMirror();
    }
    // Recursively try to link any mirrors.
    for (const auto & child : children)
    {
        success &= child->LinkAllMirrors();
    }
    return success;
}

void BaseNode::AddMirror(MirrorNode* pMirror)
{
    // Check if the mirror is already in the list.
    if (std::find(mirrors.begin(), mirrors.end(), pMirror) != mirrors.end())
        return;
    // Add mirror to list of mirrors.
    mirrors.push_front(pMirror);
}

void BaseNode::RemoveMirror(MirrorNode* pMirror)
{
    std::remove(mirrors.begin(), mirrors.end(), pMirror);
}

void BaseNode::SetNodeChanged()
{
    if (!AnyRecentChanges())
    {
        recentChanges |= 1; // Set LSB to indicate a new change.
        if (pParent != nullptr)
        {
            pParent->SetNodeChanged(); // Proceed to set parent as changed.
        }
        for(auto mirror: mirrors)
        {
            mirror->SetNodeChanged(); // Mark all the mirrors (if any) as changed.
        }
    }
}

void BaseNode::PushChangeHistory()
{
    if (AnyRecentChanges())
    {
        recentChanges &= ~1; // Clear LSB. Remove indication of most recent change.
        recentChanges += 2; // increase change count by 2 (actual count is  'recentChanges >> 2') 
        // Also push changes for all childs.
        for (const auto & child : children)
        {
            child->PushChangeHistory();
        }
    }
}
void BaseNode::ClearAllChanges()
{
    if (AnyChanges())
    {
        recentChanges = 0;      // Clear change history.
        attributeChanges = 0;   // Clear attribute changes.
                                // Do the same for all children.
        for (const auto & child : children)
        {
            child->ClearAllChanges();
        }
    }
}

// Flag attribute
void  BaseNode::SetFlags(const char* pValues)
{
    // Syntax: "[+/-]flag1|[+/-]flag2" eg. "+flag1|+flag2|-flag3" set 
    // Parse string and set flags accordingly.
    static const char flagDelimiter = '|';
    bool moreflags = true;

    if (pValues == nullptr) return;
    while (moreflags)
    {
        FlagType flag = FlagType::none;
        // End of string stop parsing.
        if (*pValues == 0) break;

        // Skip leading white spaces.
        while (*pValues != 0 && isspace((unsigned char)*pValues))
        {
            ++pValues;
        }

        bool removeFlag = false;
        if (pValues[0] == '-')
        {
            removeFlag = true;
            pValues++;
        }
        else if (pValues[0] == '+')
        {
            pValues++;
        }

        // Iterate over all flags to find a string match.    
        for (const auto &flagTuple : flagNames)
        {
            const char* pFlagName = get<0>(flagTuple).c_str();
            const char* pValue = pValues;

            while (*pValue == *pFlagName  && *pFlagName != 0 && *pValue != 0 && *pValue != flagDelimiter)
            {
                pValue++;
                pFlagName++;
            }

            // Skip trailing white spaces.
            while (*pValue != 0 && isspace((unsigned char)*pValue))
            {
                ++pValue;
            }

            // did we match to the end of flag name?
            if (*pFlagName == 0)
            {
                if (*pValue == flagDelimiter || *pValue == 0)
                {                    
                    if (*pValue == flagDelimiter) pValues = pValue + 1;
                    else moreflags = false;
                    flag = get<1>(flagTuple);
                    break;
                }
            }
        }
        if (flag != FlagType::none)
        {
            SetFlag(flag, !removeFlag);
        }
        else
        {
            return;
        }
    } // End of while loop.
}

void BaseNode::GetFlags(string &strFlags) const
{
    strFlags.clear();
    for (const auto & flagTup : flagNames)
    {
        if ((get<1>(flagTup) & nodeFlag) != FlagType::none)
        {
            if (!strFlags.empty())
            {
                strFlags.push_back('|');
            }
            strFlags.append(get<0>(flagTup));
        }
    }
}

void BaseNode::SetFlag(FlagType flag, bool value)
{
    FlagType oldFlags = nodeFlag;
    if (value)
    {
        // Add flag.
        nodeFlag |= flag;
    }
    else
    {
        // Remove flag
        nodeFlag &= (~flag);
    }
    if (oldFlags != nodeFlag)
    {
        SetAttributeChanged(flagsAttrib.GetID()); // Mark the change.;
    }
}

bool BaseNode::GetFlag(FlagType flag) const
{
    return (flag & nodeFlag) != FlagType::none;
}

// Info attribute
void BaseNode::SetInfo(const char* pValue)
{
    info = pValue;
    SetAttributeChanged(flagsAttrib.GetID()); // Mark the change.
}

// SetAttribute
// Must return true if a given attribute is handled.
bool BaseNode::SetAttribute(attribID_t attribID, const char* pAttributeValue)
{
    if(flagsAttrib.GetID() == attribID)
    {
        SetFlags(pAttributeValue);
    }
    else if(infoAttrib.GetID() == attribID)
    {
        SetInfo(pAttributeValue);
    }
    else 
    {
        return false; // Unknown attribute return false.
    }

    return true;
}

// GetAttribute
// Must return true if a given attribute is handled.
bool BaseNode::GetAttribute(attribID_t attribID, string &strAttributeValue) const
{
    //if (attribID & touchedAttributes) return false;
    if (flagsAttrib.GetID() == attribID)
    {
        GetFlags(strAttributeValue);
    }
    else if (infoAttrib.GetID() == attribID)
    {
        strAttributeValue = info;
    }
    else
    {
        return false; // Unknown attribute return false.
    }
    return true;
}

void BaseNode::Print(int indentLevel) const
{
    std::cout << std::string(indentLevel, '-') << name << "  " << typeid(*this).name() << std::endl;
    for(const auto &child : children)
    {
        child->Print(indentLevel + 1);
    }
}

// The argument nodePath contains the relative or absolute path to the node to search for.
// An absolute path is specified with a leading "/" eg. "/node1/node2/nodeToFind", in which case the function will begin the search from the root node.
// The syntax of a relative path is "node2/nodeToFind" or "../node1/node2/nodeToFind" and the search is started from this node.
// Return value: If the node is found a pointer to the requested node is returned, otherwise a nullptr is returned.
BaseNode* BaseNode::FindNodeInternal(const char * pNodePath, bool allowPartialMatch, bool resolveMirrors)
{
    if (resolveMirrors && (typeid(*this) == typeid(MirrorNode)))
    {
        MirrorNode * pMirror = dynamic_cast<MirrorNode*> (this);
        if (pMirror == nullptr || pMirror->GetMirrorSource() == nullptr)
        {
            // ToDo: Handle error.
            return nullptr;
        }
        return pMirror->GetMirrorSource()->FindNodeInternal(pNodePath, allowPartialMatch, resolveMirrors);
    }
    static const char pathDelimiter = '/';
    if (pNodePath == nullptr) return nullptr;
    // Special case for an empty string - just return this.
    if (pNodePath[0] == 0) return this;
    // Special case if first character is a pathDelimiter we should start the search from the root node.
    if (pNodePath[0] == pathDelimiter)
    {
        BaseNode* pNode = this;
        // Traverse to root.
        while (pNode->pParent)
        {
            pNode = pNode->pParent;
        }
        return pNode->FindNodeInternal(pNodePath + 1, allowPartialMatch, resolveMirrors);
    }
    // Special case if first two characters are ".." - then return the parent if not null.
    if (pNodePath[0] == '.' && pNodePath[1] != 0 && pNodePath[1] == '.' && pParent)
    {
        if (pNodePath[2] == pathDelimiter) return pParent->FindNodeInternal(pNodePath + 3, allowPartialMatch, resolveMirrors);
        if (pNodePath[2] == 0) return pParent;
    }

    // Otherwise lookup (next) node name in nodePath.
    // Otherwise search children for at string match with next node.
    BaseNode* pPartialMatchNode = nullptr;
    const char* pPartialMatchPos = nullptr;
    for (const auto &child : children)
    {
        const char* pChildName = child->name.c_str();
        const char* pNodeName = pNodePath;

        while (*pNodeName == *pChildName && *pChildName != 0 && *pNodeName != 0 && *pNodeName != pathDelimiter)
        {
            pNodeName++;
            pChildName++;
        }
        // did we match to the end of child name?
        if (*pChildName == 0)
        {
            if (*pNodeName == pathDelimiter) { return child->FindNodeInternal(pNodeName + 1, allowPartialMatch, resolveMirrors);}
            if (*pNodeName == 0) 
            {
                if (resolveMirrors && (typeid(*(child.get())) == typeid(MirrorNode)))
                {
                    MirrorNode * pMirror = dynamic_cast<MirrorNode*> (child.get());
                    if (pMirror == nullptr || pMirror->GetMirrorSource() == nullptr)
                    {
                        // ToDo: Handle error.
                        return nullptr;
                    }
                    return pMirror->GetMirrorSource();
                }
                return child.get();
            }
        }
        if (allowPartialMatch) // If partial match is allowed checke that the specified named is matched entirely.
        {
            if (*pNodeName == pathDelimiter || *pNodeName == 0)
            {
                if (pPartialMatchNode) { return nullptr; } // ambiguity - More than one child node have a partial match.
                pPartialMatchNode = child.get();
                pPartialMatchPos = pNodeName;
            }
        }
    }
    if (allowPartialMatch && pPartialMatchNode != nullptr)
    {
        if (*pPartialMatchPos == 0) 
        { 
            if (resolveMirrors && (typeid(*pPartialMatchNode) == typeid(MirrorNode)))
            {
                MirrorNode * pMirror = dynamic_cast<MirrorNode*> (pPartialMatchNode);
                if (pMirror == nullptr || pMirror->GetMirrorSource() == nullptr)
                {
                    // ToDo: Handle error.
                    return nullptr;
                }
                return pMirror->GetMirrorSource();
            }
            return pPartialMatchNode; 
        }
        return pPartialMatchNode->FindNodeInternal(pPartialMatchPos + 1, allowPartialMatch, resolveMirrors); // (+1) It must have been a delimiter. Move one pos forward.
    }
    return nullptr;
}


void BaseNode::GetOptions(std::string& strOptions, const std::vector<std::string> & options, const char delimiter)
{
    strOptions.clear();
    bool first = true;
    for (auto opt : options)
    {
        if (!first)
        {
            strOptions.push_back(delimiter);
        }
        strOptions += opt;
        first = false;
    }
}
void BaseNode::SetOptions(const char* pValues, std::vector<std::string> & options, const char delimiter)
{
    options.clear();
    istringstream streamValues(pValues);
    string opt;
    while (getline(streamValues, opt, delimiter))
    {
        options.push_back(opt);
    }
}

void BaseNode::SetOptionValue(const char* pValue, const std::vector<std::string> & options, int &value)
{
    int i = 0;
    for (auto opt : options)
    {
        if (strcasecmp(pValue, opt.c_str()) == 0)
        {
            value = i;
            return;
        }
        i++;
    }
    value = -1;
}

File: ROBINinspect/nodes/BaseNode.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ROBINinspect
{
    public class BaseNode
    {
        protected String name;
        public String Name
        {
            get { return name; }
        }

        protected List<BaseNode> children = new List<BaseNode>();
        public IReadOnlyCollection<BaseNode> Children
        {
            get { return children.AsReadOnly(); }
        }

        public IEnumerable<BaseNode> ChildrenIterator()
        {
            int numChilds = children.Count;
            foreach(BaseNode ownChild in children)
            { 
                yield return ownChild;
            }
            if (this.GetType() == typeof(MirrorNode))
            {
                
                foreach (BaseNode mirrorChild in (this as MirrorNode).MirrorSource.children)
                { 
                    yield return mirrorChild; 
                }
            }
        }


        protected BaseNode parentNode;
        public BaseNode ParentNode
        {
            get { return parentNode; }
        }

        public BaseNode(string nodeName, BaseNode parent = null)
        {
            name = nodeName;
            parentNode = parent;
            if (parentNode != null)
                parentNode.SetNodeChanged();
        }

        public BaseNode GetRoot()
        {
            BaseNode node = this;
            // Traverse to root.
            while (node.ParentNode != null)
            {
                node = node.ParentNode;
            }
            return node;
        }

        #region Attributes
        #region Flag attribute
        private FlagType flags;
        public FlagType Flags
        {
            get { return flags; }
            set
            {
                flags = value;
                SetAttributeChanged(AttributeTypes.flags);
            }
        }
        public void SetFlags(String strFlags)
        {
            // Syntax: "[+/-]flag1|[+/-]flag2" eg. "+flag1|+flag2|-flag3" set 
            // Parse string and set flags accordingly.
            const char flagDelimiter = '|';
            string[] arrFlags = strFlags.Split(flagDelimiter);
            foreach (String strFlag in arrFlags)
            {
                // Skip leading white spaces.
                String tmp = strFlag.Trim();
                bool removeFlag = false;
                if (strFlag[0] == '-')
                {
                    removeFlag = true;
                    tmp  = tmp.TrimStart('-');
                }
                else if (strFlag[0] == '+')
                {
                    tmp = tmp.Trim('+');
                }
                foreach (FlagType flag in Enum.GetValues(typeof(FlagType)))
                {
                    if (tmp.Equals(flag.ToString()))
                    {
                        SetFlag(flag, !removeFlag);
                    }
                }
            }
        }

        public void SetFlag(FlagType flag, bool value, bool recursive = false)
        {
            FlagType oldFlags = flags;
            if (value)
            {   // Add flag.
                flags |= flag;
            }
            else
            {   // Remove flag
                flags &= (~flag);
            }
            if (oldFlags != flags)
            {
                SetAttributeChanged(AttributeTypes.flags); // Mark the change.;
            }
            if (recursive)
            {
                foreach (BaseNode child in children)
                {
                    child.SetFlag(flag, value, recursive);
                }
            }
        }
        void GetFlags(ref string strFlags)
        {
            strFlags = String.Empty;
            foreach (FlagType flag in Enum.GetValues(typeof(FlagType)))
            {
                if ((flag & flags) != FlagType.none)
                {
                    if (!(strFlags.Length == 0))
                    {
                        strFlags += "|";
                    }
                    strFlags += flag.ToString();
                }
            }
        }
        #endregion

        private String info;
        public String Info
        {
            get { return info; }
            set
            {
                info = value;
                SetAttributeChanged(AttributeTypes.info);
            }
        }

        // -------- Attributes --------
        // Get the string representation of a given attribute.
        public virtual bool GetAttribute(AttributeTypes attribID, ref String strAttributeValue)
        {
            switch (attribID)
            {
                case AttributeTypes.flags:
                    GetFlags(ref strAttributeValue);
                    break;
                case AttributeTypes.info:
                    strAttributeValue = info;
                    break;
                default:
                    return false; // Unknown attribute return false.
            }
            return true;
        }

        // Sets the given attribute, based on the string argument.
        // Inherited nodes should call and return a base implementation if it does not handle the attributeID.
        // Returns true is the attribute is handled. 
        public virtual bool SetAttribute(AttributeTypes attribID, String strAttributeValue)
        {
            switch (attribID)
            {
                case AttributeTypes.flags:
                    SetFlags(strAttributeValue);
                    break;
                case AttributeTypes.info:
                    Info = strAttributeValue;
                    break;
                default:
                    return false; // Unknown attribute return false.
            }
            return true;
        }

        private UInt32 attributeChanges;
        private UInt32 recentChanges;
        private UInt32 touchedAttributes;
        // Mark a given attribute as changed.
        public void SetAttributeChanged(AttributeTypes attribID)
        {
            if (attributeChanges == 0)
            {   // Notify the node tree that a change has been applied.
               SetNodeChanged();
            }
            attributeChanges |= (1U << (int)attribID);
            touchedAttributes |= attributeChanges;
        }
        public virtual bool IsAttributeChanged(AttributeTypes attribID) { return (attributeChanges & (1 << (int)attribID)) != 0; }
        public virtual bool IsAttributeUsed(AttributeTypes attribID) { return (touchedAttributes & (1 << (int)attribID)) != 0; }

        void SetNodeChanged()
        {
            if (!AnyRecentChanges())
            {
                recentChanges |= 1; // Set LSB to indicate a new change.
                if (parentNode != null)
                {
                    parentNode.SetNodeChanged(); // Proceed to set parent as changed.
                }
                //for (auto mirror: mirrors)
                //{
                //    mirror->SetNodeChanged(); // Mark all the mirrors (if any) as changed.
                //}
            }
        }
        #endregion

        public bool AnyChanges()  { return recentChanges != 0; }
        public bool AnyRecentChanges() { return (recentChanges & 1) == 1; }
        public int RecentChangeCount() { return (int)((recentChanges >> 2) + (recentChanges & 1)); }

        public void AddChild(BaseNode childNode)
        {
            if (Children.Contains(childNode)) return; // No duplicates allowed.
            // ToDo: Validate node name (must be unique and no special characters which are not allowed in an XML tag name)
            // Problem regarding mirrors: it's impossible to ensure a unique node name before mirros are linked.
            children.Add(childNode);
            childNode.parentNode = this;
            SetNodeChanged();
        }

        public BaseNode FindNodeInChildren(String nodeName)
        {
            //if (children == null) return null;
            foreach (BaseNode node in ChildrenIterator())
            {
                if (node.name.Equals(nodeName)) return node;
            }
            // Not found return null.
            return null;
        }

        public BaseNode FindNode(String nodePath, bool allowPartialMatch = false, bool resolveMirrors = true)
        {
            return FindNodeInternal(nodePath, 0, allowPartialMatch, resolveMirrors);
        }
        // The argument nodePath contains the relative or absolute path to the node to search for.
        // An absolute path is specified with a leading "/" eg. "/node1/node2/nodeToFind", in which case the function will begin the search from the root node.
        // The syntax of a relative path is "node2/nodeToFind" or "../node1/node2/nodeToFind" and the search is started from this node.
        // Return value: If the node is found a pointer to the requested node is returned, otherwise null is returned.
        private BaseNode FindNodeInternal(String nodePath, int strPos, bool allowPartialMatch , bool resolveMirrors)
        {
            if (resolveMirrors && (this.GetType() == typeof(MirrorNode)))
            {
                MirrorNode mirror = this as MirrorNode;
                if (mirror == null || mirror.MirrorSource == null)
                {
                    // ToDo: Handle error.
                    return null;
                }
                return mirror.MirrorSource.FindNodeInternal(nodePath, strPos, allowPartialMatch, resolveMirrors);
            }
            const char pathDelimiter = '/';
            if (nodePath == null) return null;
            // Special case for an empty string - just return this.
            if (nodePath.Length <= strPos) return this;
            // Special case if first character is a pathDelimiter we should start the search from the root node.
            if (strPos == 0 && nodePath[0] == pathDelimiter)
            {
                BaseNode node = GetRoot();
                return node.FindNodeInternal(nodePath, 1, allowPartialMatch, resolveMirrors);
            }
            // Special case if first two characters are ".." - then return the parent if not null.
            if (nodePath[strPos] == '.' && strPos <= nodePath.Length && nodePath[strPos + 1] == '.' && ParentNode != null)
            {
                if (strPos + 2 == nodePath.Length) return ParentNode;
                if (nodePath[strPos + 2] == pathDelimiter) return parentNode.FindNodeInternal(nodePath, strPos + 3, allowPartialMatch, resolveMirrors);
            }
            // Otherwise lookup (next) node name in nodePath.
            int sepPos = nodePath.IndexOf(pathDelimiter, strPos);
            if (sepPos == -1)
            {
                sepPos = nodePath.Length;
            }
            string nodeName = nodePath.Substring(strPos, sepPos - strPos);
            // Otherwise search children for at string match with next node.
            BaseNode partialMatchNode = null;
            foreach(BaseNode child in  children)
            {
                if(allowPartialMatch)
                {
                    if (child.name.StartsWith(nodeName))
                    {
                        if (partialMatchNode != null)
                        {  // ambiguity - More than one child node have a partial match.
                            return null;
                        }
                        partialMatchNode = child;
                    }
                }
                else
                {
                    if (child.name.Equals(nodeName))
                    {
                        return child.FindNodeInternal(nodePath, sepPos+1, allowPartialMatch, resolveMirrors);
                    }
                }
            }

            if (allowPartialMatch && partialMatchNode != null)
            {
                if (sepPos >= nodePath.Length)
                {
                    if (resolveMirrors && (partialMatchNode.GetType() == typeof(MirrorNode)))
                    {
                        MirrorNode mirror = partialMatchNode as MirrorNode;
                        if (mirror == null || mirror.MirrorSource == null)
                        {
                            // ToDo: Handle error.
                            return null;
                        }
                        return mirror.MirrorSource;
                    }
                    return partialMatchNode;
                }
                return partialMatchNode.FindNodeInternal(nodePath, sepPos + 1, allowPartialMatch, resolveMirrors); // (+1) It must have been a delimiter. Move one pos forward.
            }
            return null;
        }

        #region Mirrors
        List<MirrorNode> mirrors = new List<MirrorNode>();
        public void AddMirror(MirrorNode mirror)
        {
            // Check if the mirror is already in the list.
            if (mirrors.Contains(mirror))
                return;
            // Add mirror to list of mirrors.
            mirrors.Add(mirror);
        }

        public void RemoveMirror(MirrorNode mirror)
        {
            mirrors.Remove(mirror);
        }

        private bool LinkAllMirrors()
        {
            bool success = true;
            MirrorNode mirror = this as MirrorNode;
            if (mirror != null) // Check if this node is a mirror.
            {
                success &= mirror.LinkMirror();
            }
            // Recursively try to link any mirrors.
            foreach(BaseNode child in children)
            {
                success &= child.LinkAllMirrors();
            }
            return success;
        }


        #endregion

    }
}
